Justfile

```diff
@@ -0,0 +1,84 @@
+# FIXME: some of these tests don't work very well on MacOS at the moment. Should fix
+# them at earliest convenience :)
+# Also in addition to this, the `nextest-integration` test is currently disabled on macos
+with_macos_excludes := if os() == "macos" {
+    "--exclude node-runtime --exclude runtime-params-estimator --exclude near-network --exclude estimator-warehouse"
+} else {
+    ""
+}
+nightly_flags := "--features nightly,test_features"
+bash := if os() == "macos" { "/bin/bash" } else { "/usr/bin/bash" }
+
+# all the tests, as close to CI as possible
+test: test-ci test-extra
+
+# only the tests that are exactly the same as the ones in CI
+test-ci: (nextest "stable") (nextest "nightly") check-style
+
+# tests that are as close to CI as possible, but not exactly the same code
+test-extra: check-lychee
+
+# all cargo tests, TYPE is "stable" or "nightly"
+nextest TYPE: (nextest-unit TYPE) (nextest-integration TYPE)
```

Comment:
It is important that we allow passing filters and other arbitrary flags through to `nextest` as well… Local development will usually not run the full test suite every time, rather filtering for a subset of tests they are interested in with flags like `-p` and just plain test names.

Reply:
Hmm… I do agree that it would be good to have this, but I’m not sure `Justfile` would be the right tool for it? Basically, this would be going away from the behaviour of CI, and so is probably less important.

And TBH I don’t think it makes sense for people to learn the syntax for our specific `Justfile` when they already know the syntax for `cargo`. So I’m thinking of `just` as just a way to reproduce CI behavior locally, and nothing more.

WDYT?

Reply:
> And TBH I don’t think it makes sense for people to learn the syntax for our specific Justfile when they already know the syntax for cargo. So I’m thinking of just as just a way to reproduce CI behavior locally, and nothing more.

How does this align with the intent to move the `style` checks out from within nextest and the hope that we eventually have a single command local development experience, not multiple?

Like, if Justfile’s singular concern is reproducing the CI, then it is not all that appealing in majority of local development workflows (e.g. I’d imagine rarely anybody runs the entire test suite in every local iteration cycle) which then makes it a poor replacement for the current `style` tests in nextest too…?

Reply:
Hmm… I don’t think the current `style` tests are any better than a `just check-style` command would be? If being run with `-p near-vm-runner`, then the style tests still won’t be run.

In my mind it’s a balance to find between "re-implementing the world behind a new CLI" (the solution of offering toggles for everything, though a reasonable amount of toggles, like a `CARGOFLAGS` environment variable, might make sense) and "having a single command to run all of CI in a monolithic way".

So basically my goal here is to have many commands to run each part of CI, more or less configurable depending on what’s easily doable, and then people can mix-and-match them depending on their needs. TBH I think I personally will mostly want to just run the `just` command that’s being run by whichever CI job I see failing, and just use regular `cargo nextest -p whichevercrateImworkingon` usually 😅 

WDYT?

Reply:
Hmm actually… maybe just having `just packages=near-vm-runner,node-runtime test` run all the test suite (including style checks) but only with `-p near-vm-runtime -p node-runtime` would be enough to address your concerns?

It’d be strictly better than the statu quo, where style checks don’t run when nextest is run with `-p`, and it wouldn’t require too large an additional API surface :)

WDYT?

Reply:
Good point! This is now done, though using `just nextest stable -p near-vm-runner` will have the maybe-unexpected result of also running tests for the `integration-tests` package :)

Reply:
I didn’t extend it to all of `just test`, because I think there’d be a risk of misunderstanding to which sub-command the additional flags would apply; so it currently applies only to `nextest`, `nextest-unit` and `nextest-integration`.